// conf/config.mjs
export const arxliveCopy =
	'search-datavis-arxlive-copy-n6ltva3lqh7x7xb6ucpaqfb5a4.eu-west-2.es.amazonaws.com';
export const ec2 = 'http://ec2-3-8-167-48.eu-west-2.compute.amazonaws.com';
export const spotlightEndpoint = `${ec2}:2222/rest/annotate`;
export const confidenceValues = [
	0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1,
];

export const settings = {
	snapshotSettings: {
		region: 'eu-west-2',
		bucketName: 'nesta-datavis-arxlive-copy-snapshot',
		snapshotRole: 'arxliveSnapshotRole',
		repository: 'spotlight-annotation-backups',
		awsID: '195787726158',
	},
};

export const defaultMapping = {
	type: 'nested',
	properties: {
		URI: {
			type: 'keyword',
		},
		confidence: {
			type: 'integer',
		},
		percentageOfSecondRank: {
			type: 'float',
		},
		similarityScore: {
			type: 'float',
		},
		surfaceForm: {
			type: 'text',
		},
		duplicates_60: {
			type: 'integer',
		},
		duplicates_10: {
			type: 'integer',
		}
	},
};

export const metaDataMapping = {
	properties: {
		confidence_avg: {
			type: 'float'
		},
		confidence_max: {
			type: 'integer'
		},
		confidence_min: {
			type: 'integer'
		},
		entities_count: {
			type: 'integer'
		},
		dupes_10_ratio: {
			type: 'float'
		},
		dupes_60_ratio: {
			type: 'float'
		},
		dupes_10_count: {
			type: 'integer'
		},
		dupes_60_count: {
			type: 'integer'
		},
		confidence_counts: {
			properties: {
				"0": {
					type: 'integer'
				},
				"10": {
					type: 'integer'
				},
				"20": {
					type: 'integer'
				},
				"30": {
					type: 'integer'
				},
				"40": {
					type: 'integer'
				},
				"50": {
					type: 'integer'
				},
				"60": {
					type: 'integer'
				},
				"70": {
					type: 'integer'
				},
				"80": {
					type: 'integer'
				},
				"90": {
					type: 'integer'
				},
				"100": {
					type: 'integer'
				}
			}
		}
	}

}
