// es/requests.mjs
import { defaultProvider } from '@aws-sdk/credential-provider-node';
import { SignatureV4 } from '@aws-sdk/signature-v4';
import { NodeHttpHandler } from '@aws-sdk/node-http-handler';
import sha256 from '@aws-crypto/sha256-browser';
import { HttpRequest } from '@aws-sdk/protocol-http';
const { Sha256 } = sha256;

import { logger } from 'logging/logging.mjs';

const signer = new SignatureV4({
	credentials: defaultProvider(),
	region: 'eu-west-2',
	service: 'es',
	sha256: Sha256,
});

/**
 * @function buildRequest
 * @description builds a HttpRequest object using the AWS sdk. Needed for siginin the request using Environment variables.
 * @param {string} domain ElasticSearch domain on which to make request.
 * @param {string} path - additional path, appended after the domain in the request URL.
 * @param {string} method - HTTP request method (GET, POST, etc.).
 * @param {Object} [options]
 * @param {Object|string} [options.payload] - optional payload for the request. Can be passed as object and subsequently stringifyed.
 * @param {Object} [options.query={}] - optionaly query object for using the search API.
 * @returns {HttpRequest} the AWS HttpRequest object, signed using AWS credentials.
 */
export const buildRequest = (
	domain,
	path,
	method,
	{ payload, query = {} } = {}
) => {
	const body =
		payload && typeof payload !== 'string'
			? JSON.stringify(payload)
			: payload;
	return new HttpRequest({
		body,
		method,
		path,
		query,
		headers: {
			'Content-Type': 'application/json',
			host: domain,
		},
		hostname: domain,
	});
};

/**
 * @function makeRequest
 * @description makes a request using a HttpRequest object.
 * @param {HttpRequest} request - the HttpRequest object built using {@link buildRequest}
 * @param {Object} [options={}]
 * @param {boolean} [options.verbose] - whether to log the output of the request and response.
 * @returns {Object} the HttpResponse object
 */
export const makeRequest = async (request, { verbose = false } = {}) => {
	// Sign the request
	const signedRequest = await signer.sign(request);
	// Send the request
	const client = new NodeHttpHandler();
	const { response } = await client.handle(signedRequest);
	const responseBody = await parseResponseBody(response);
	return {
		code: response.statusCode,
		message: response.body.statusMessage,
		body: responseBody,
	};
};

/**
 * @function parseResponseBody
 * @description helper function which returns promise of signed response object's body
 * @param {Object} response - response object obtained from {@link makeRequest}
 * @returns {Promise} promise which resolves to the response's body 
 */
const parseResponseBody = (response) => {
	let responseBody = '';
	return new Promise((resolve, _) => {
		response.body.on('data', (chunk) => {
			responseBody += chunk;
		});
		response.body.on('end', () => {
			resolve(JSON.parse(responseBody));
		});
	});
}
