// dbpedia/spotlight.mjs
import fetch from 'undici';

import { logger } from '../logging.mjs';
import { spotlightEndpoint } from 'conf/config.mjs'

export const annotate = async (
	text,
	confidence,
	{ endpoint = spotlightEndpoint } = {}
) => {
	const url = new URL(endpoint);
	const body = `text=${text}&confidence=${confidence}`;
	const response = await fetch(url, {
		method: 'POST',
		headers: {
			Accept: 'application/json',
			'content-type': 'application/x-www-form-urlencoded',
		},
		body: encodeURI(body),
	});
	if (response.ok) {
		return response.json();
	} else {
		logger.warn(response);
		logger.error(new Error('Something went wrong with the annotation...'));
	}
};

export const parseAnnotationResults = (results) => {
	const resultSubset = results.Resources
		? results.Resources.map((result) => {
				return {
					'@URI': result['@URI'],
					'@surfaceForm': result['@surfaceForm'],
					'@similarityScore': result['@similarityScore'],
					'@percentageOfSecondRank':
						result['@percentageOfSecondRank'],
				};
		  })
		: [];

	return {
		confidence: results['@confidence'],
		results: resultSubset,
	};
};
